### PR TITLE
Add more cluster cache test.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -181,8 +184,12 @@ func (r *RegionsInfo) addRegion(region *metapb.Region) {
 		key:    searchKey(encodeRegionEndKey(region)),
 		region: region,
 	}
-	// TODO: if duplicated, panic!!!
-	r.searchRegions.ReplaceOrInsert(item)
+
+	oldItem := r.searchRegions.ReplaceOrInsert(item)
+	if oldItem != nil && bytes.Equal(oldItem.(*searchKeyItem).key, item.key) {
```
Comment (siddontang) [inline server/cache.go]: should we need to check the key here?

File: server/cache.go [hunk 2]
```diff
@@ -181,8 +184,12 @@ func (r *RegionsInfo) addRegion(region *metapb.Region) {
...
+		log.Fatalf("addRegion for already existed region - %v", region)
+	}
+
 	r.regions[region.GetId()] = region
```
Comment (siddontang) [inline server/cache.go]: I think we should check here too.

**Comment by siddontang** [comment]
I only see check, no any tests?

**Comment by qiuyesuifeng** [comment]
@disksing  PTAL

File: server/cluster_worker_test.go [hunk 3]
```diff
@@ -308,6 +308,21 @@ func mustGetRegion(c *C, cluster *raftCluster, key []byte, expect *metapb.Region
...
+
+	for _, key := range keys {
+		mockRegion := &metapb.Region{EndKey: key}
+		item := &searchKeyItem{
+			key: searchKey(encodeRegionEndKey(mockRegion)),
+		}
+
+		getItem := cacheRegions.searchRegions.Get(item)
+		c.Assert(getItem.(*searchKeyItem).key, DeepEquals, item.key)
+	}
```
Comment (disksing) [inline server/cluster_worker_test.go]: I think it's better to iterate `searchRegions` and compare each item with the expect keys, then we can be sure there is no extra items in the btree.

Comment (qiuyesuifeng) [inline server/cluster_worker_test.go]: Line 313 has checked the length of cache, so i think it is enough here.

Comment (disksing) [inline server/cluster_worker_test.go]: Yeah. It should be enough. LGTM.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**